Continuous integration with GitHub Actions

nchaimov opened the pull request:
Adds continuous integration with GitHub Actions.

The current implementation defines two containers, one for building Spindle with `--with-rm=serial` and one for `--with-rm=flux`. The Flux container uses Docker Compose to set up a cluster of 4 nodes. The containers are set up such that when building the image, the Spindle repo is copied into the container and built there. The GitHub Actions workflow checks out the commit to be tested and builds the images incorporating the checked-out commit.

The tests will fail until #97 and #100 are merged, as these fix bugs affecting the Ubuntu environment in the containers. See https://github.com/ParaToolsInc/spindle-testing/actions/runs/17871033110 for an example in a fork which already has these fixes applied to see the tests passing. And see https://github.com/ParaToolsInc/Spindle/actions/runs/17871526378 for what happens with the commits without these fixes, showing that failing tests indeed fail the CI run as expected.

If we want to require manual authorization before running the tests, we will have to first create an environment in GitHub and set the workflow to run in that environment.

nchaimov committed:
Continuous integration with GitHub Actions

nchaimov commented:
With #97 and #100 merged, I've rebased off devel. The tests now pass, as can be seen in the CI run on my fork: https://github.com/ParaToolsInc/Spindle/actions/runs/17885768345

nchaimov commented:
The CI workflow is now configured to run in the `Spindle CI` environment, which will require authorization before running the CI workflow using the organization's resources. Forks can create an environment with the same name to run using the fork owner's resources. Additionally, I have verified that the CI workflow can be run locally by using [gh-act](https://github.com/nektos/gh-act).

I think this should be ready to merge. The only possibly remaining issue is when to run the CI workflow. As currently configured, it will run on all pull requests, and any commits to any branch in this repo. That will generate a request for review on every commit, but if we don't want that, we could instead run only on pull requests, or only on commits to specific branches, such as to `devel`.

mplegendre commented:
I'm adding @alecbcs as a reviewer. He's a developer at here and an expert on how we manage CI. 
Alec, help/guidance would be appreciated here. And we can talk more about this tomorrow if you have questions.

alecbcs commented on .github/workflows/CI.yml:
```diff
@@ -0,0 +1,97 @@
+name: CI
+
+on: [push, pull_request, workflow_dispatch]
```
I'd suggest limiting the execution of CI to the `devel` branch (and to PRs that target devel) which will reduce the number of parallel executions that can occur when you push to a branch and then create a PR. Additionally helps reduce the load on the limited number of Actions Runners we all share in the LLNL org.

https://github.com/LLNL/Spindle/blob/d5e582f316e03f23afcf6f410b06566ebade5f3c/.github/workflows/ci.yml#L2-L8

nchaimov replied:
I will do this, thanks

alecbcs commented on .github/workflows/CI.yml:
```diff
@@ -0,0 +1,97 @@
+    runs-on: ubuntu-latest
+    timeout-minutes: 20
+    steps:
+      - name: Check out Spindle
+        uses: actions/checkout@v5
```
To minimize the chance of external modifications creating a failure state in CI, I typically recommend that GitHub projects lock their actions to a specific commit and then use GitHub Dependabot to automatically bump that commit periodically.

https://github.com/LLNL/Spindle/blob/d5e582f316e03f23afcf6f410b06566ebade5f3c/.github/workflows/style.yml#L12

https://github.com/LLNL/Spindle/blob/d5e582f316e03f23afcf6f410b06566ebade5f3c/.github/dependabot.yml#L1-L6

alecbcs commented on .github/workflows/CI.yml:
```diff
@@ -0,0 +1,97 @@
+    steps:
+      - name: Check out Spindle
+        uses: actions/checkout@v5
+        with:
+          path: 'Spindle'
```
```suggestion
```

Did you all decide to clone into `Spindle/` for a particular use case? Normally we recommend to clone without a path which clone into the CI/CD runner current working directory. (And then minimizing the number of cd hierarchies you need later on.)

nchaimov replied:
I don't think this is needed for our use cases. I can get rid of it.

alecbcs reviewed (commented):
Hi @nchaimov! Nice job on the CI/CD pipeline! A couple quick suggestions and otherwise this looks good to me. 

- Tiny knitpick to follow common standards I'd suggest changing `CI.yml` -> `ci.yml`

I submitted a simple example CI/CD workflow in #106 to show the best practices I typically build off of which allow for filtering based on files changed, and break the workflows out into their own separate files.

nchaimov commented:
I've made the following changes: 

- renamed from `CI.yml` to `ci.yml`
- removed checkout path
- use specific commit for actions
- use Dependabot to update actions
- only run on commits or pull requests to `devel`
- added `permissions` and `concurrency` sections from #106

mplegendre commented:
I think this is looking great. Anything else before I merge?

nchaimov commented:
It's ready from my perspective

alecbcs reviewed (approved):
Looks good to me. Thanks @nchaimov for the quick turnaround!
